x-pack/plugin/enrich/src/main/java/org/elasticsearch/xpack/enrich/EnrichPolicyRunner.java

```diff
@@ -109,19 +114,74 @@ private void validateMappings(final GetIndexResponse getIndexResponse) {
         logger.debug("Policy [{}]: Validating [{}] source mappings", policyName, sourceIndices);
         for (String sourceIndex : sourceIndices) {
             Map<String, Object> mapping = getMappings(getIndexResponse, sourceIndex);
-            Map<?, ?> properties = ((Map<?, ?>) mapping.get("properties"));
-            if (properties == null) {
-                listener.onFailure(
-                    new ElasticsearchException(
-                        "Enrich policy execution for [{}] failed. Could not read mapping for source [{}] included by pattern [{}]",
-                        policyName, sourceIndex, policy.getIndices()));
+            // First ensure mapping is set
+            if (mapping.get("properties") == null) {
+                throw new ElasticsearchException(
+                    "Enrich policy execution for [{}] failed. Could not read mapping for source [{}] included by pattern [{}]",
+                    policyName, sourceIndex, policy.getIndices());
+            }
+            // Validate the key and values
+            try {
+                validateField(mapping, policy.getEnrichKey(), true);
+                for (String valueFieldName : policy.getEnrichValues()) {
+                    validateField(mapping, valueFieldName, false);
+                }
+            } catch (ElasticsearchException e) {
+                throw new ElasticsearchException(
+                        "Enrich policy execution for [{}] failed while validating field mappings for index [{}]",
+                        e, policyName, sourceIndex);
+            }
+        }
+    }
+
+    private void validateField(Map<?, ?> properties, String fieldName, boolean fieldRequired) {
+        assert fieldName != null && !fieldName.isEmpty() : "Field name cannot be null or empty";
+        String[] fieldParts = fieldName.split("\\.");
+        StringBuilder parent = new StringBuilder();
+        Map<?, ?> currentField = properties;
+        boolean onRoot = true;
+        for (String fieldPart : fieldParts) {
+            // Ensure that the current field is of object type only (not a nested type or a non compound field)
+            Object type = currentField.get("type");
+            if (type != null) {
+                throw new ElasticsearchException(
```

Comment:
an object field can be specifically configured with: `type: object`, so we need to check for this here